Interpolation for arbitrary grids

jleinonen opened the pull request:
<!-- markdownlint-disable MD013-->
# Earth2Studio Pull Request

## Description
<!-- Provide a standalone description of changes in this PR. -->
<!-- Reference any issues closed by this PR with "closes #1234". -->
<!-- Note: The pull request title will be included in the CHANGELOG. -->

* Add interpolation between arbitrary 2D lat-lon grids as `earth2studio.utils.interp.LatLonInterpolation`. This is implemented as a `torch.nn.Module` because initializing the mapping between the grids can be a bit slow but after that the interpolation of the values is very fast.
* Move the interpolation function from the CorrDiff implementation to `earth2studio.utils.interp` for easier reuse in e.g. other CorrDiff implementations.
* Add test for `LatLonInterpolation`.

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/earth2studio/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/earth2studio/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/earth2studio/issues) is linked to this pull request.

## Dependencies

<!-- Call out any new dependencies needed if any -->

jleinonen commented:
/blossom-ci

NickGeneva commented on earth2studio/utils/interp.py:
```diff
@@ -0,0 +1,170 @@
+
+    Parameters
+    ----------
+    lat0 : torch.Tensor [H_in, W_in]
```
Update param names

jleinonen replied:
Changed in https://github.com/NVIDIA/earth2studio/pull/159/commits/cc59471fae79f5305c9e0cb512ca4d56e0f76834.

NickGeneva commented on test/utils/test_interp.py:
```diff
@@ -0,0 +1,38 @@
+
+@pytest.mark.parametrize("device", ["cpu", "cuda:0"])
+@pytest.mark.parametrize("input_type", ["zeros", "random", "gradient"])
+def test_interpolation(device, input_type):
```
Any change you can implement at unit test here that checks the qualitative value of the output.

So have maybe you have a 2x2 reference in a [0,1] box with some fixed values at the corners. Then you have a few points at like [0.25,0.25] [0.25,0.5] [0.5.0.25] and [0.5 0.5].

And confirm the outputs are analytically what one would expect

jleinonen replied:
Added in https://github.com/NVIDIA/earth2studio/pull/159/commits/4df40a1f4ca4dba0c8da60f9d9fc6a6544546754.

NickGeneva commented on earth2studio/utils/interp.py:
```diff
@@ -0,0 +1,170 @@
+from torch import Tensor, nn
+
+
+def latlon_interpolation_regular(
```
Confirming this is just a 1:1 copy from corrdiff yeah?

jleinonen replied:
Yes, it's just cut/pasted over from corrdiff.

NickGeneva commented on earth2studio/utils/interp.py:
```diff
@@ -0,0 +1,170 @@
+        self.register_buffer("i_map", i_map)
+        self.register_buffer("j_map", j_map)
+
+    @torch.inference_mode()
```
Can you try adding @torch.compile here

jleinonen replied:
Added in https://github.com/NVIDIA/earth2studio/pull/159/commits/4df40a1f4ca4dba0c8da60f9d9fc6a6544546754, seems to work.

NickGeneva commented on earth2studio/utils/interp.py:
```diff
@@ -0,0 +1,170 @@
+        ----------
+        values : torch.Tensor [..., H_in, W_in]
+            Input values defined over (lat0, lon0) that will be interpolated onto
+            (lat1, lon1) grid.
```
Can you add a return doc here so it tells the expected dim

Returns
-------
torch.Tensor
    Interpolated tensor of shape [..., W_out, H_out]

jleinonen replied:
Changed in https://github.com/NVIDIA/earth2studio/pull/159/commits/cc59471fae79f5305c9e0cb512ca4d56e0f76834.

NickGeneva commented on earth2studio/utils/interp.py:
```diff
@@ -0,0 +1,170 @@
+
+        Parameters
+        ----------
+        values : torch.Tensor [..., H_in, W_in]
```
Nit:

Move the dims ([..., H_in, W_in]) into the description line

jleinonen replied:
Changed in https://github.com/NVIDIA/earth2studio/pull/159/commits/4df40a1f4ca4dba0c8da60f9d9fc6a6544546754.

NickGeneva commented on earth2studio/utils/interp.py:
```diff
@@ -0,0 +1,170 @@
+        Tensor of input latitude coordinates
+    lon0 : torch.Tensor [H_in, W_in]
+        Tensor of input longitude coordinates
+    lat1 : torch.Tensor [W_out, H_out]
```
Same Nit as below, dims down into description. And update to torch.Tensor | np.ndarray

jleinonen replied:
Changed in https://github.com/NVIDIA/earth2studio/pull/159/commits/4df40a1f4ca4dba0c8da60f9d9fc6a6544546754.

NickGeneva commented on earth2studio/utils/interp.py:
```diff
@@ -0,0 +1,170 @@
+    def __init__(
+        self,
+        lat_in: Tensor | ArrayLike,
+        lon_in: Tensor | ArrayLike,
```
Nit: typically like to use torch.Tensor here just so its explicit what tensor/package its expecting

jleinonen replied:
Changed in https://github.com/NVIDIA/earth2studio/pull/159/commits/cc59471fae79f5305c9e0cb512ca4d56e0f76834.

NickGeneva commented:
Can you also add this to the API docs: https://github.com/NVIDIA/earth2studio/blob/main/docs/modules/utils.rst#modearth2studioutils-coordinates

Under the coordinate section under the utils, can add:

```
.. autosummary::
   :toctree: generated/utils/
   :template: class.rst

 utils.interp.LatLonInterpolation
```

You can make just the API docs with:
```
pip install -e .[docs]
make docs
cd docs/_build/html
```

NickGeneva commented:
Lastly add a note to the change log that this was added!

NickGeneva commented:
Just some minor nits / edits and some doc items. Looks great to me otherwise, thanks Jussi

NickGeneva commented:
/blossom-ci

jleinonen commented:
> Can you also add this to the API docs: https://github.com/NVIDIA/earth2studio/blob/main/docs/modules/utils.rst#modearth2studioutils-coordinates
> 
> Under the coordinate section under the utils, can add:
> 
> ```
> .. autosummary::
>    :toctree: generated/utils/
>    :template: class.rst
> 
>  utils.interp.LatLonInterpolation
> ```
> 
> You can make just the API docs with:
> 
> ```
> pip install -e .[docs]
> make docs
> cd docs/_build/html
> ```

Added in https://github.com/NVIDIA/earth2studio/pull/159/commits/e0a7395fe498a3225d269a2b0aaa8c6f0d8ac445, could you check that it's correct?

> Lastly add a note to the change log that this was added!

Also added in the above commit.

jleinonen commented:
/blossom-ci

NickGeneva commented:
/blossom-ci

NickGeneva commented:
/blossom-ci

NickGeneva commented on test/utils/test_interp.py:
```diff
@@ -0,0 +1,66 @@
+import numpy as np
```
Ooops this file needs a license header. Thanks!

NickGeneva commented:
/blossom-ci
